### ISSUE TITLE
Can't force reconnect while within transaction and connection lost 

### ISSUE THREAD

**Issue by halaei** [laravel/framework#15865]
- Laravel Version: 5.3.*

### Description:
It seems that the framework don't provide any way to reconnect to database while within a transaction and connection is lost.

### Steps To Reproduce:

It is hard to reproduce. Not having the option of handling such a situation is bad. It may happen if someone is working with daemon queue workers and for some reason a transaction is not closed within a job and is leaked out. It may also happen for queries that are expected to take long.

```php
    public function reconnect($force = false)
    {
        ...
    }
    ...
    public function setPdo($pdo, $force = false)
    {
        if ($this->transactions >= 1 && ! $force) {
            throw new RuntimeException("Can't swap PDO instance while within transaction.");
        }

        $this->transactions = 0;

        $this->pdo = $pdo;

        return $this;
    }
````

**Comment by themsaid** [comment]
what do you suggest? how do you think we should handle this?

**Comment by halaei** [comment]
I was working on a PR to add a $force = true param to DB::reconnect().

**Comment by taylorotwell** [comment]
There is another PR already open to fix this I believe? https://github.com/laravel/framework/pull/15511

**Comment by halaei** [comment]
#15511 sounds to be different. My PR lets the client code do this:

```php
try {
DB::beginTransaction();
...
//possibly nested transactions are here without any extra try-catch
...
DB::commit();
} catch (Exception $e) {
   try {
   while(DB::transactionLevel() > 0) DB::rollBack();
   } catch (Exception $e) {
       //This looks only possible after the pull request: to handle lost connection or any other reason that makes rollbacks impossible.
       // Without this PR the application must be aborted, a terrible situation for daemon workers as an example.
       DB::reconnect(true);
       throw $e;
   }
}
```

**Comment by themsaid** [comment]
@halaei was this taken care of by any of your PRs?

**Comment by halaei** [comment]
Does github have a problem with replies via email? I answered 2ice via email but both are failed I think?
Sorry for possible duplicate replies:
@themsaid Yes check this please: https://github.com/laravel/framework/pull/15931
It is simple and hopefully working. This issue is now closed for 5.3. I may not try for a proper change for 5.1 though.

**Comment by halaei** [comment]
Yes check this please: https://github.com/laravel/framework/pull/15931
It is simple and hopefully working. This issue is now closed for 5.3. I may
not try for a proper change for 5.1 though.
